src/sage/combinat/abstract_tree.py

```diff
@@ -639,15 +638,127 @@ def action(x):
                 # subtrees, and should not be exploded again, but instead
                 # should be manipulated and removed from the stack.
                 stack.append(None)
-                for i in range(len(node)):
-                    subtree = node[-i - 1]
+                for subtree in reversed(node):
                     if not subtree.is_empty():
                         stack.append(subtree)
             else:
                 stack.pop()
                 node = stack.pop()
                 action(node)
 
+    def contour_traversal(self,first_action=None, middle_action=None, final_action=None, leaf_action=[]):
+        r"""
+        Run the counterclockwise countour traversal algorithm (iterative
+        implementation) and subject every node encountered
+        to some procedure ``first_action``, ``middle_action`` or ``final_action`` each time it reaches it. 
+        The algorithm is::
+
+            manipulate the root with function `first_action`;
+            iteratively manipulate the root with function `middle_action` 
+                and explore each subtree (by the algorithm) from the
+                leftmost one to the rightmost one;
+            then manipulate the root with function `final_action`;
+            if the root is a leaf it only manipulate with function `leaf_action`.
```

Comment:
```suggestion

        ALGORITHM:

        -  apply `first_action` to the root
        - iteratively apply `middle_action` to the root and traverse each subtree from the
          leftmost one to the rightmost one
        - apply `final_action` to the root if it is not a leaf, and apply `leaf_action` to the root otherwise
```
(I'm not sure whether my modification introduce mistakes, please double check!)

Reply:
```suggestion

        ALGORITHM:
        
        - if the root is a leaf, apply `leaf_action`
        - else
            -  apply `first_action` to the root
            - iteratively apply `middle_action` to the root and traverse each subtree from the
          leftmost one to the rightmost one
            - apply `final_action` to the root
```

Reply:
It does. This is correct.